source/lib.js

```diff
@@ -6,6 +6,50 @@
  * here eventually.
  */
 
+/**
+ * Duplicate a block and attach statements from the clause to be prefixed to the block.
+ * Adds braces if the block is bare.
+ *
+ * @returns the duplicated block with prefix statements attached.
+ */
+function blockWithPrefix(clause, block) {
+  if (!clause) return block
+  const { blockPrefix } = clause
+
+  if (blockPrefix && blockPrefix.length) {
+    const indent = getIndent(block.expressions[0])
+    // Match prefix statements to block indent level
+    const prefixStatements = indent
+      ? blockPrefix.map((statement) => [indent, ...statement.slice(1)])
+      : blockPrefix
+
+    const expressions = [...prefixStatements, ...block.expressions]
+
+    block = {
+      ...block,
+      expressions,
+      children: block.children === block.expressions ? expressions :
+        block.children.map((c) => c === block.expressions ? expressions : c),
+    }
+    // Add braces if block lacked them
+    if (block.bare) {
+      // Now copied, so mutation is OK
+      block.children = [[" {\n", indent], ...block.children, "}"]
+      block.bare = false
+    }
+  }
+
+  return block
+}
+
+function closest(node, types) {
+  do {
+    if (types.includes(node.type)) {
+      return node
+    }
+  } while (node = node.parent)
+}
```

Comment:
This is a special case of [findAncestor](https://github.com/DanielXMoore/Civet/blob/9b28e3a351d130ce031c7b28b50bcb6d2a499315/source/parser.hera#L7376). I'm fine with a separate implementation that's more efficient, but perhaps we could name it similarly (`findAncestorOfType` might be more descriptive), and maybe move `findAncestor`'s code to be next to this in lib?

Reply:
I named this one after the https://developer.mozilla.org/en-US/docs/Web/API/Element/closest API and eventually want to add full query selector support. `findAncestorOfType` is a better description of what it is doing currently, however.